Backport changes from 1.21.1

IThundxr opened the pull request.

IThundxr committed:
Backport changes from 1.21.1

IThundxr committed:
fix

IThundxr committed:
Fix building

IThundxr committed:
fix compile error

IThundxr committed:
fix

IThundxr committed:
fix build for real

PepperCode1 commented on buildSrc/src/main/kotlin/dev/engine_room/gradle/transitive/TransitiveSourceSetsExtension.kt:
```diff
@@ -25,7 +26,7 @@ open class TransitiveSourceSetsExtension(private val project: Project) {
     fun createCompileConfigurations() {
         val configs = transitives.mapValues { (sourceSet, _) ->
-            project.configurations.create("for${sourceSet.name.capitalize()}") {
+            project.configurations.create("for${sourceSet.name.replaceFirstChar { it.uppercase() }}") {
```
Why was this change made?

IThundxr replied:
capitalize is deprecated, it shouldn't be used anymore

PepperCode1 commented on common/src/main/java/dev/engine_room/flywheel/impl/compat/SodiumCompat.java:
```diff
@@ -0,0 +1,19 @@
+import net.minecraft.world.level.block.entity.BlockEntityType;
+
+public class SodiumCompat {
+	public static <T extends BlockEntity> Object forBlockEntityType(BlockEntityType<T> type) {
```
Rename to `addPredicate` for consistency with the other method.

PepperCode1 commented on common/src/main/java/dev/engine_room/flywheel/impl/mixin/BlockEntityTypeMixin.java:
```diff
@@ -5,15 +5,22 @@
 @Mixin(BlockEntityType.class)
 abstract class BlockEntityTypeMixin<T extends BlockEntity> implements BlockEntityTypeExtension<T> {
 	@Unique
+	@Nullable
```
Also add `@Nullable` to the equivalent field in `EntityTypeMixin`.

PepperCode1 commented on forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java:
```diff
@@ -63,6 +65,9 @@ private static void clientInit(IEventBus forgeEventBus, IEventBus modEventBus) {
 		CrashReportCallables.registerCrashCallable("Flywheel Backend", BackendManagerImpl::getBackendString);
 		FlwImpl.init();
+
+		if (CompatMods.EMBEDDIUM.isLoaded())
```
Use braces.

PepperCode1 commented on forge/src/main/resources/META-INF/mods.toml:
```diff
@@ -26,10 +26,16 @@ mandatory = true
 versionRange = "${forge_version_range}"
 side = "CLIENT"
 
+[[dependencies.flywheel]]
```
Use `${mod_id}` instead of `flywheel`.

PepperCode1 reviewed (changes requested):
After addressing review comments, also organize imports in all changed files as some are out of order.

IThundxr committed:
address reviews

Jozufozu commented on forge/src/main/java/dev/engine_room/flywheel/impl/mixin/sodium/ChunkBuilderMeshingTaskMixin.java:
Don't we still need these on 1.20.1? Is sodium 0.6 getting backported?

IThundxr replied:
Right...

IThundxr replied:
It's not needed on forge as there's proper embeddium compat, but on fabric it's still needed and it needs to be controlled by the version of sodium that is loaded

IThundxr committed:
Fix sodium compat

IThundxr committed:
address requested changes

IThundxr committed:
mark rubidium as incompatible

IThundxr committed:
add missed call

Jozufozu committed:
Should have worn steel toe boots

- Add "stub" sourceset to each subproject
- Directly pass vararg sourcesets to methods in PlatformExtension to
  avoid automatically shipping jars with the api stubs
- We may have to include stubs in setupLoomMod, but I don't think so
- A lot of this can be stripped back out if we don't need stub sources
  for the forge/fabric subprojects

PepperCode1 committed:
Guarded stubs

- Add Sodium 0.6 and Iris API stubs to stubs source set and remove Gradle dependencies on local Sodium jar, Iris, and Oculus
- Ensure usage of APIs that may not exist at runtime is in private classes and access is always guarded
- Change ShadersModHandler
  - Rename to ShadersModHelper
  - Convert methods to check for Iris' and Optifine's presence into static final fields
  - Move implementation to impl source set in form of IrisCompat and OptifineCompat classes
- Rename CompatMods to CompatMod and add public field to access mod ID
- Set BlockEntityType's Sodium predicate to null after it is removed
- Update repository links
- Remove local libs repository
